src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/WhenShouldBeUsedCheck.java

```diff
@@ -0,0 +1,159 @@
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code when.should.be.used}
+ * </li>
+ * </ul>
+ *
+ * @since 10.18.0
+ */
+
+@StatelessCheck
+public class WhenShouldBeUsedCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "when.should.be.used";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {TokenTypes.LITERAL_CASE};
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final DetailAST statementList = getStatementListOfCase(ast);
+        final boolean hasSingleIfStatement = statementList != null
+                                                && hasSingleIfStatementInCaseBlock(statementList);
+        final boolean hasPatternLabel = hasPatternLabel(ast);
+        if (hasPatternLabel && hasSingleIfStatement) {
+            log(ast, MSG_KEY);
+        }
+
+    }
+
+    /**
+     * Get the statement list token of the case block.
+     *
+     * @param caseAST the AST node representing {@code LITERAL_CASE}
+     * @return the AST node representing {@code SLIST} of the current case
+     */
+    private static DetailAST getStatementListOfCase(DetailAST caseAST) {
+        final DetailAST caseParent = caseAST.getParent();
+        final DetailAST statementList = caseParent.findFirstToken(TokenTypes.SLIST);
+
+        return Optional.ofNullable(statementList)
+               .map(ast -> ast.findFirstToken(TokenTypes.SLIST))
+               .orElse(statementList);
+    }
+
+    /**
+     * Check if the case block has a pattern variable definition
+     * or a record pattern definition.
+     *
+     * @param caseAST the AST node representing {@code LITERAL_CASE}
+     * @return true if the case block has a pattern label, false otherwise
+     */
+    private static boolean hasPatternLabel(DetailAST caseAST) {
+        return caseAST.findFirstToken(TokenTypes.PATTERN_VARIABLE_DEF) != null
+                || caseAST.findFirstToken(TokenTypes.RECORD_PATTERN_DEF) != null
+                || caseAST.findFirstToken(TokenTypes.PATTERN_DEF) != null;
+    }
+
+    /**
+     * Check if the case block has a single if statement.
+     *
+     * @param statementList the AST node representing {@code SLIST} of the current case
+     * @return true if the case block has a single if statement, false otherwise
+     */
+    private static boolean hasSingleIfStatementInCaseBlock(DetailAST statementList) {
+        final int[] acceptableStatementsInCaseBlock = {
+            TokenTypes.LITERAL_BREAK,
+            TokenTypes.EMPTY_STAT,
+            TokenTypes.RCURLY,
+        };
```

Comment:
Please move this to a static final field.

Reply:
done